NetCoreMentoring/src/NetCoreMentoring.App/Infrastructure/ImageCacheMiddleware.cs

```diff
@@ -0,0 +1,80 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Configuration;
+using NetCoreMentoring.Core;
+using NetCoreMentoring.Core.Utilities;
+
+namespace NetCoreMentoring.App.Infrastructure
+{
+    // I use caching with help of attribute instead
+    // Don't delete this just for completing homework module
+    public class ImageCacheMiddleware
+    {
+        private readonly RequestDelegate next;
+        private static readonly Dictionary<string, string> SupportedPictureContentTypes = new Dictionary<string, string>()
+        {
+            {"image/jpeg", "jpeg"},
+            {"image/png", "png"}
+        };
+        private readonly string CacheImagePath;
+        private readonly IConfiguration _configuration;
+
+        public ImageCacheMiddleware(RequestDelegate next, IConfiguration configuration)
+        {
+            this.next = next;
+            _configuration = configuration;
+
+            CacheImagePath = configuration["CacheImagePath"];
+        }
+
+        public async Task Invoke(HttpContext context)
+        {
+            // We can't read response body in middleware, problem described here:
+            // https://stackoverflow.com/questions/43403941/how-to-read-asp-net-core-response-body/43404745
+
+            var originalBody = context.Response.Body;
+            try
+            {
+                await using var memStream = new MemoryStream();
+                context.Response.Body = memStream;
+
+                await next(context);
+
+                memStream.Position = 0;
+                await memStream.CopyToAsync(originalBody);
+
+                if (SupportedPictureContentTypes.Keys.Contains(context.Response.ContentType ?? ""))
+                {
+                    if (Directory.GetFiles(CacheImagePath).Length >=
+                        int.Parse(CacheImagePath)) return;
+
+                    if (!context.Request.Query.TryGetValue("categoryId", out var categoryId)) return;
+
+                    // Delete previously cached image for this category
+                    var allCachedImages = Directory.GetFiles(_configuration["CacheImagePath"]);
```

Comment:
This code is a but hard to read, at least consider splitting it into helper methods

Reply:
done